_posts/2024-03-03-pubsub-consumers.md

```diff
@@ -0,0 +1,57 @@
+
+Note: With MQTT protocol version 5, MQTT introduced the concept of subscriber groups. Subscriber groups are a feature that addresses the limitations of generic fanout messaging on scaled subscribers but also isn't supported by every MQTT broker.
+
+## Publish/Subscribe with queues
+
+Message queues are another concept used by many messaging technologies that extend the publish-subscribe model with an explicit entity on the broker that can be shared by subscribers. This allows clients to use a "competing consumers" pattern where a message in a queue is delivered to only one client instance. Therefore a queue can be assigned to a logical client and client instances all compete for messages on the same queue to ensure that the message is processed once independently from the number of clients available. This makes client scaling simple and very flexible as clients can be scaled up or down at any point and messages can be distributed easily across all active client instances.
+
+
+![](/assets/amqp-clients.jpg)
+In this diagram, each logical application has its own message queue. Using the "competing consumers" model, exactly one client instance is able to consumer a message per logical application.
+
+
+One downside of this approach is that it makes broadcasts more difficult. If there are some messages that should be received by all clients, each client requires an additional, dedicated queue (this is the reason NServiceBus doesn't support broadcast events and instead recommends other approaches).
+
+## Publish/Subscribe with logs
+
+Event Stream Processing is another approach to messaging that has gained a lot of popularity. Kafka is probably the best example, but both AWS and Azure offer similar products as a service. I'm using Kafka as the technology name here, as it's so commonly used.
+
+Kafka publishers store messages in a topic partition. Every topic has one or more partitions which persistes the messages. Subscribers can then read messages from a partition at any point without the message being removed. Kafka also defines the concept of client groups, which ultimately is the same as our "logical client". Each client group can have multiple client instances that belong to the group.
+
+In kafka, a partition can only be read by one client instance within the same client group. This has several implications:
+* Similar to "traditional" messages queues, one event will be processed by a single client instance within the logical application.
+* It's harder to mix broadcast and competing consumers models at the same time because a client instance belongs to a single client group and broadcasts effectly only work across client groups, not across client-instances.
```

Comment:
```suggestion
* It's harder to mix broadcast and competing consumer models at the same time because a client instance belongs to a single client group, and broadcasts effectively only work across client groups, not across client instances.
```